language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/ContractTypeCompanion.scala

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.client.binding
+
+import com.daml.ledger.api.refinements.ApiTypes.TemplateId
+import com.daml.ledger.api.v1.{value => rpcvalue}
+import encoding.ExerciseOn
+
+import annotation.nowarn
+
+/** Common superclass of template and interface companions objects. */
+abstract class ContractTypeCompanion[T] extends ValueRefCompanion {
+
+  /** Alias for contract IDs for this template or interface. Can be used
+    * interchangeably with its expansion.
+    */
+  type ContractId = Primitive.ContractId[T]
+
+  val id: Primitive.TemplateId[T]
+
+  override protected lazy val ` dataTypeId` = TemplateId.unwrap(id)
```

Comment:
I'm not familiar with this naming style, what is the benefit of using a leading space in the identifier?

Reply:
To avoid name collision with generated code. `dataTypeId` is a legitimate Daml-LF identifier, but this symbol is not.